sympy/integrals/heurisch.py

```diff
@@ -450,7 +450,6 @@ def heurisch(f, x, rewrite=False, hints=None, mappings=None, retries=3,
     rev_mapping = {v: k for k, v in mapping}                            #
     if mappings is None:                                                #
         # optimizing the number of permutations of mapping              #
-        assert mapping[-1][0] == x  # if not, find it and correct this comment
```

Comment:
I don't think this should be simply deleted. The next line should probably be run only if this is true.

Reply:
It does not work if i do this. Actually, i do not understand the working of heurisch, i just tried tweaking something.

Reply:
I don't see why `ordered` should always give x last. Maybe change the code to find x instead of assuming it is last.

Reply:
It seems that `x` is the only symbol that may be included in `terms` and, therefore, it tends to be last in reverse order. The issue with a `BaseScalar` variable is that it is not included in `terms` by `components` as it is not a `Symbol`:

    if x in f.free_symbols:
        if f.is_Symbol:
            result.add(f)

This could probably be fixed by changing the condition to `f.is_symbol` and adding `is_symbol = True` to `BaseScalar`. To prevent a `MatrixSymbol` from passing the test, `and f.is_commutative` could also be added.

Reply:
Thanks for the help.